File: src/pages/code-tool/components/template-unified-extractor/template_extractor.rs
use regex::Regex;
use std::collections::HashSet;

/// Vue 模板全能提取器 - Rust 实现
/// 字段结构与 Vue 组件及 Python 版本保持高度一致
pub struct TemplateExtractor {
    pub mustache_props: Vec<String>,
    pub mustache_methods: Vec<String>,
    pub event_functions: Vec<String>,
    pub colon_props: Vec<String>,

    // 预编译正则以提高性能
    re_mustache: Regex,
    re_event: Regex,
    re_colon: Regex,
}

impl TemplateExtractor {
    pub fn new() -> Self {
        Self {
            mustache_props: Vec::new(),
            mustache_methods: Vec::new(),
            event_functions: Vec::new(),
            colon_props: Vec::new(),
            re_mustache: Regex::new(r"\{\{(.*?)\}\}").unwrap(),
            re_event: Regex::new(r#"@([\w-]+)((?:\.[\w-]+)*)\s*=\s*["']([^"']+)["']"#).unwrap(),
            re_colon: Regex::new(r#":([\w.-]+)="([^"]+)""#).unwrap(),
        }
    }

    fn extract_base_variable(&self, expression: &str) -> String {
        expression.trim()
            .split('.')
            .next()
            .unwrap_or("")
            .split('[')
            .next()
            .unwrap_or("")
            .to_string()
    }

    fn parse_vue_handler(&self, expression: &str) -> Option<String> {
        let content = expression.trim();
        let forbidden = ['=', '+', '-', '*', '/', '&', '|', '!', '>'];
        if content.is_empty() || content.contains(|c| forbidden.contains(&c)) {
            return None;
        }

        // 提取首个合法标识符
        content.split(|c: char| !c.is_alphanumeric() && c != '_' && c != '$')
            .next()
            .map(|s| s.to_string())
    }

    pub fn process_all(&mut self, raw_template: &str) {
        self.reset_result();
        if raw_template.is_empty() { return; }

        // 1. Mustache
        let mut m_props = HashSet::new();
        let mut m_methods = HashSet::new();
        for cap in self.re_mustache.captures_iter(raw_template) {
            let inner = cap[1].trim();
            if inner.contains('(') {
                if let Some(name) = inner.split('(').next() {
                    m_methods.insert(name.trim().to_string());
                }
            } else {
                let base = self.extract_base_variable(inner);
                if !base.is_empty() { m_props.insert(base); }
            }
        }
        self.mustache_props = m_props.into_iter().collect();
        self.mustache_methods = m_methods.into_iter().collect();

        // 2. Events
        let mut e_funcs = HashSet::new();
        for cap in self.re_event.captures_iter(raw_template) {
            if let Some(func) = self.parse_vue_handler(&cap[3]) {
                e_funcs.insert(func);
            }
        }
        self.event_functions = e_funcs.into_iter().collect();

        // 3. Colon Props
        let mut c_props = HashSet::new();
        for cap in self.re_colon.captures_iter(raw_template) {
            let base = self.extract_base_variable(&cap[2]);
            if !base.is_empty() { c_props.insert(base); }
        }
        self.colon_props = c_props.into_iter().collect();
    }

    pub fn reset_result(&mut self) {
        self.mustache_props.clear();
        self.mustache_methods.clear();
        self.event_functions.clear();
        self.colon_props.clear();
    }
}

File: project/code-tool/pages/code-tool/components/template-unified-extractor/template_extractor.py
import re
from typing import List, Set, Optional

class TemplateExtractor:
    """
    Vue 模板全能提取器 - Python 实现
    保持与 template-unified-extractor.vue 逻辑一致
    """

    MUSTACHE_REGEX = r'\{\{(.*?)\}\}'
    VUE_EVENT_REGEX = r'@([\w-]+)((?:\.[\w-]+)*)\s*=\s*["\']([^"\']+)["\']'
    COLON_PROP_REGEX = r':([\w.-]+)="([^"]+)"'

    def __init__(self):
        self.mustache_props: List[str] = []
        self.mustache_methods: List[str] = []
        self.event_functions: List[str] = []
        self.colon_props: List[str] = []

    def _extract_base_variable(self, expression: str) -> str:
        # 模拟 project-common.js 中的 extract_base_variable
        # 提取基础变量名 (例如: item.name -> item)
        content = expression.strip()
        return content.split('.')[0].split('[')[0].strip()

    def _parse_vue_handler(self, expression: str) -> Optional[str]:
        content = expression.strip()
        # 排除复杂表达式
        if not content or any(op in content for op in "=+-*/&|!>"):
            return None
        # 提取函数名
        match = re.match(r'^([a-zA-Z_$][\w$]*)', content)
        return match.group(1) if match else None

    def process_all(self, raw_template: str):
        if not raw_template:
            self.reset_result()
            return

        # 1. 提取 Mustache {{ }}
        m_props: Set[str] = set()
        m_methods: Set[str] = set()
        for match in re.finditer(self.MUSTACHE_REGEX, raw_template):
            inner = match.group(1).strip()
            if '(' in inner:
                func_name = inner.split('(')[0].strip()
                if func_name:
                    m_methods.add(func_name)
            else:
                base_prop = self._extract_base_variable(inner)
                if base_prop:
                    m_props.add(base_prop)

        self.mustache_props = sorted(list(m_props))
        self.mustache_methods = sorted(list(m_methods))

        # 2. 提取事件 @
        e_funcs: Set[str] = set()
        for match in re.finditer(self.VUE_EVENT_REGEX, raw_template):
            func_name = self._parse_vue_handler(match.group(3))
            if func_name:
                e_funcs.add(func_name)
        self.event_functions = sorted(list(e_funcs))

        # 3. 提取绑定属性 :
        c_props: Set[str] = set()
        for match in re.finditer(self.COLON_PROP_REGEX, raw_template):
            attr_val = match.group(2).strip()
            if attr_val:
                base_val = self._extract_base_variable(attr_val)
                c_props.add(base_val)
        self.colon_props = sorted(list(c_props))

    def reset_result(self):
        self.mustache_props = []
        self.mustache_methods = []
        self.event_functions = []
        self.colon_props = []

    def get_summary(self):
        return {
            "interpolation_count": len(self.mustache_props) + len(self.mustache_methods),
            "event_count": len(self.event_functions),
            "prop_count": len(self.colon_props)
        }
